File: regions/system/library/go/building-blocks/pubsub.go
package buildingblocks

import (
	"context"
	"time"
)

// Message represents a pub/sub message.
type Message struct {
	Topic     string            `json:"topic"`
	Data      []byte            `json:"data"`
	Metadata  map[string]string `json:"metadata,omitempty"`
	ID        string            `json:"id"`
	Timestamp time.Time         `json:"timestamp"`
}

// PubSub provides publish/subscribe messaging capabilities.
type PubSub interface {
	Component
	Publish(ctx context.Context, msg *Message) error
	Subscribe(ctx context.Context, topic string) (<-chan *Message, error)
}

File: regions/system/library/typescript/building-blocks/src/pubsub.ts
import type { Component } from './component.js';

export interface Message {
  topic: string;
  data: Uint8Array;
  metadata: Record<string, string>;
  id: string;
}

export interface MessageHandler {
  handle(message: Message): Promise<void>;
}

export interface PubSub extends Component {
  publish(topic: string, data: Uint8Array, metadata?: Record<string, string>): Promise<void>;
  subscribe(topic: string, handler: MessageHandler): Promise<string>;
  unsubscribe(subscriptionId: string): Promise<void>;
}
